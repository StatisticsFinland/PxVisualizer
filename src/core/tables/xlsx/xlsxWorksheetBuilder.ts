import { worksheetProperties } from "./xlsxPropertyConstants";
import { TCell } from "./xlsxTypes";
import { escapeXmlFunctionChars } from "./xlsxUtils";

export const buildXMLSheet = (rows: TCell[][]) : string => {
    const xmlColumns = buildColumns(Math.max(...rows.map(r => r.length)));
    const xmlRows = buildRows(rows);
    return wrapInSheetTemplate(xmlColumns, xmlRows);
};

// Columns format from: https://msdn.microsoft.com/en-us/library/office/documentformat.openxml.spreadsheet.columns
// Column format from: https://msdn.microsoft.com/en-us/library/office/documentformat.openxml.spreadsheet.column
function buildColumns(numOfcols: number) : string
{
    const cols: string[] = [];
    for(let i = 1; i <= numOfcols; i++ ) {
        cols.push(`<col min="${i}" max="${i}" width="20"/>`)
    }
    return cols ? `<cols>${cols}</cols>` : '';
};

// Row format from: https://learn.microsoft.com/en-us/dotnet/api/documentformat.openxml.spreadsheet.row
function buildRows(rows: TCell[][])
{
    return rows.map((row, rowIndex) => {
        const rowCells = row
            .map((cell, cellIndex) => buildCell(cell, cellIndex, rowIndex + 1))
            .filter(cell => cell !== null)
            .join('');
        return `<row r="${rowIndex + 1}">${rowCells}</row>`;
    }).join('');
}

// Cell format from: https://learn.microsoft.com/en-us/dotnet/api/documentformat.openxml.spreadsheet.cell
function buildCell(cell: TCell, columnNumber: number, rowNumber: number) : string | null
{    
    if (cell === null) return null;

    return (
        typeof(cell) === 'string'
            ? `<c r="${getColumnLetter(columnNumber)}${rowNumber}" t="inlineStr"><is><t>${escapeXmlFunctionChars(cell)}</t></is></c>`
        : `<c r="$${getColumnLetter(columnNumber)}${rowNumber}"><v>${cell}</v></c>`
    );
}

function getColumnLetter(colIndex: number) : string {  
    
    // If we run out of letters, we combine new letters after the prevous "prefix" ie. AA, AB, AC etc.
    const prefix = Math.floor(colIndex / 26);
    const letter = String.fromCharCode(97 + (colIndex % 26)).toUpperCase();
    
    if (prefix === 0) return letter;
    else return getColumnLetter(prefix - 1) + letter;
};

const wrapInSheetTemplate = (xmlColumns: string, xmlRows: string) => `<?xml version="1.0" ?>
<worksheet ${worksheetProperties.join(' ')}>
    ${xmlColumns}
    <sheetData>
        ${xmlRows}
    </sheetData>
</worksheet>`;