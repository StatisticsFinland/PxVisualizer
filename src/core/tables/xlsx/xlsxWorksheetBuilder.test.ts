import { XLSX_FIXTURES_COMPLEX_VIEW_ROW_AND_COL_VARS, XLSX_FIXTURES_ONE_CELL_ESCAPE_CHARACTERS } from "../fixtures/xlsxFixtures";
import { buildXMLSheet } from "./xlsxWorksheetBuilder";

describe('Worksheet xml generation tests', () => {
    it('Should generate expexted xml output with the provided fixture with fi locale', () => {
        const sheet : string = buildXMLSheet(XLSX_FIXTURES_COMPLEX_VIEW_ROW_AND_COL_VARS['fi']);
        expect(sheet).toMatchSnapshot();
    });

    it('Should generate expexted xml output with the provided fixture with sv locale', () => {
        const sheet : string = buildXMLSheet(XLSX_FIXTURES_COMPLEX_VIEW_ROW_AND_COL_VARS['sv']);
        expect(sheet).toMatchSnapshot();
    });

    it('Should generate expexted xml output with the provided fixture with en locale', () => {
        const sheet : string = buildXMLSheet(XLSX_FIXTURES_COMPLEX_VIEW_ROW_AND_COL_VARS['en']);
        expect(sheet).toMatchSnapshot();
    });

    it('Should generate expected xml output with provided fixture with en locale including special characters', () => {
        const sheet: string = buildXMLSheet(XLSX_FIXTURES_ONE_CELL_ESCAPE_CHARACTERS['en']);
        expect(sheet).toMatchSnapshot();
    });
});
