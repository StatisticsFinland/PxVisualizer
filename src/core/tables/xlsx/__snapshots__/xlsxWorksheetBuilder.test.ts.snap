// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Worksheet xml generation tests Should generate expexted xml output with the provided fixture with en locale 1`] = `
"<?xml version="1.0" ?>
<worksheet xmlns="http://schemas.openxmlformats.org/spreadsheetml/2006/main" xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006" xmlns:mv="urn:schemas-microsoft-com:mac:vml" xmlns:mx="http://schemas.microsoft.com/office/mac/excel/2008/main" xmlns:r="http://schemas.openxmlformats.org/officeDocument/2006/relationships" xmlns:x14="http://schemas.microsoft.com/office/spreadsheetml/2009/9/main" xmlns:x14ac="http://schemas.microsoft.com/office/spreadsheetml/2009/9/ac" xmlns:xm="http://schemas.microsoft.com/office/excel/2006/main">
    <cols><col min="1" max="1" width="20"/>,<col min="2" max="2" width="20"/>,<col min="3" max="3" width="20"/>,<col min="4" max="4" width="20"/>,<col min="5" max="5" width="20"/>,<col min="6" max="6" width="20"/></cols>
    <sheetData>
        <row r="1"><c r="A1" t="inlineStr"><is><t>Rents per square meter (eur/m2) 2015Q1-2015Q3 by Region, Number of rooms, Type of funding</t></is></c></row><row r="2"><c r="C2" t="inlineStr"><is><t>One-room flat</t></is></c><c r="D2" t="inlineStr"><is><t>One-room flat</t></is></c><c r="E2" t="inlineStr"><is><t>Two-room flat</t></is></c><c r="F2" t="inlineStr"><is><t>Two-room flat</t></is></c></row><row r="3"><c r="C3" t="inlineStr"><is><t>Non subsidised</t></is></c><c r="D3" t="inlineStr"><is><t>Government subsidised</t></is></c><c r="E3" t="inlineStr"><is><t>Non subsidised</t></is></c><c r="F3" t="inlineStr"><is><t>Government subsidised</t></is></c></row><row r="4"><c r="C4" t="inlineStr"><is><t>Rents per square meter (eur/m2)</t></is></c><c r="D4" t="inlineStr"><is><t>Rents per square meter (eur/m2)</t></is></c><c r="E4" t="inlineStr"><is><t>Rents per square meter (eur/m2)</t></is></c><c r="F4" t="inlineStr"><is><t>Rents per square meter (eur/m2)</t></is></c></row><row r="5"><c r="A5" t="inlineStr"><is><t>2015Q1</t></is></c><c r="B5" t="inlineStr"><is><t>Helsinki</t></is></c><c r="C5"><v>22.68</v></c><c r="D5"><v>13.8</v></c><c r="E5"><v>15.77</v></c><c r="F5"><v>12.09</v></c></row><row r="6"><c r="A6" t="inlineStr"><is><t>2015Q1</t></is></c><c r="B6" t="inlineStr"><is><t>Turku</t></is></c><c r="C6"><v>15.15</v></c><c r="D6"><v>12.7</v></c><c r="E6"><v>10.5</v></c><c r="F6"><v>10.95</v></c></row><row r="7"><c r="A7" t="inlineStr"><is><t>2015Q2</t></is></c><c r="B7" t="inlineStr"><is><t>Helsinki</t></is></c><c r="C7"><v>22.87</v></c><c r="D7"><v>13.95</v></c><c r="E7"><v>15.95</v></c><c r="F7"><v>12.23</v></c></row><row r="8"><c r="A8" t="inlineStr"><is><t>2015Q2</t></is></c><c r="B8" t="inlineStr"><is><t>Turku</t></is></c><c r="C8"><v>15.18</v></c><c r="D8"><v>12.84</v></c><c r="E8"><v>10.57</v></c><c r="F8"><v>11.04</v></c></row><row r="9"><c r="A9" t="inlineStr"><is><t>2015Q3</t></is></c><c r="B9" t="inlineStr"><is><t>Helsinki</t></is></c><c r="C9"><v>23.1</v></c><c r="D9"><v>14</v></c><c r="E9"><v>16.12</v></c><c r="F9"><v>12.31</v></c></row><row r="10"><c r="A10" t="inlineStr"><is><t>2015Q3</t></is></c><c r="B10" t="inlineStr"><is><t>Turku</t></is></c><c r="C10"><v>15.32</v></c><c r="D10"><v>12.93</v></c><c r="E10"><v>10.65</v></c><c r="F10"><v>11.1</v></c></row><row r="11"><c r="A11" t="inlineStr"><is><t>Unit: eur / m2</t></is></c></row><row r="12"><c r="A12" t="inlineStr"><is><t>Source: PxVisualizer-en</t></is></c></row>
    </sheetData>
</worksheet>"
`;

exports[`Worksheet xml generation tests Should generate expexted xml output with the provided fixture with fi locale 1`] = `
"<?xml version="1.0" ?>
<worksheet xmlns="http://schemas.openxmlformats.org/spreadsheetml/2006/main" xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006" xmlns:mv="urn:schemas-microsoft-com:mac:vml" xmlns:mx="http://schemas.microsoft.com/office/mac/excel/2008/main" xmlns:r="http://schemas.openxmlformats.org/officeDocument/2006/relationships" xmlns:x14="http://schemas.microsoft.com/office/spreadsheetml/2009/9/main" xmlns:x14ac="http://schemas.microsoft.com/office/spreadsheetml/2009/9/ac" xmlns:xm="http://schemas.microsoft.com/office/excel/2006/main">
    <cols><col min="1" max="1" width="20"/>,<col min="2" max="2" width="20"/>,<col min="3" max="3" width="20"/>,<col min="4" max="4" width="20"/>,<col min="5" max="5" width="20"/>,<col min="6" max="6" width="20"/></cols>
    <sheetData>
        <row r="1"><c r="A1" t="inlineStr"><is><t>Neliövuokra (eur/m2) 2015Q1-2015Q3 muuttujina Alue, Huoneluku, Rahoitusmuoto</t></is></c></row><row r="2"><c r="C2" t="inlineStr"><is><t>Yksiöt</t></is></c><c r="D2" t="inlineStr"><is><t>Yksiöt</t></is></c><c r="E2" t="inlineStr"><is><t>Kaksiot</t></is></c><c r="F2" t="inlineStr"><is><t>Kaksiot</t></is></c></row><row r="3"><c r="C3" t="inlineStr"><is><t>Vapaarahoitteinen</t></is></c><c r="D3" t="inlineStr"><is><t>ARA</t></is></c><c r="E3" t="inlineStr"><is><t>Vapaarahoitteinen</t></is></c><c r="F3" t="inlineStr"><is><t>ARA</t></is></c></row><row r="4"><c r="C4" t="inlineStr"><is><t>Neliövuokra (eur/m2)</t></is></c><c r="D4" t="inlineStr"><is><t>Neliövuokra (eur/m2)</t></is></c><c r="E4" t="inlineStr"><is><t>Neliövuokra (eur/m2)</t></is></c><c r="F4" t="inlineStr"><is><t>Neliövuokra (eur/m2)</t></is></c></row><row r="5"><c r="A5" t="inlineStr"><is><t>2015Q1</t></is></c><c r="B5" t="inlineStr"><is><t>Helsinki</t></is></c><c r="C5"><v>22.68</v></c><c r="D5"><v>13.8</v></c><c r="E5"><v>15.77</v></c><c r="F5"><v>12.09</v></c></row><row r="6"><c r="A6" t="inlineStr"><is><t>2015Q1</t></is></c><c r="B6" t="inlineStr"><is><t>Turku</t></is></c><c r="C6"><v>15.15</v></c><c r="D6"><v>12.7</v></c><c r="E6"><v>10.5</v></c><c r="F6"><v>10.95</v></c></row><row r="7"><c r="A7" t="inlineStr"><is><t>2015Q2</t></is></c><c r="B7" t="inlineStr"><is><t>Helsinki</t></is></c><c r="C7"><v>22.87</v></c><c r="D7"><v>13.95</v></c><c r="E7"><v>15.95</v></c><c r="F7"><v>12.23</v></c></row><row r="8"><c r="A8" t="inlineStr"><is><t>2015Q2</t></is></c><c r="B8" t="inlineStr"><is><t>Turku</t></is></c><c r="C8"><v>15.18</v></c><c r="D8"><v>12.84</v></c><c r="E8"><v>10.57</v></c><c r="F8"><v>11.04</v></c></row><row r="9"><c r="A9" t="inlineStr"><is><t>2015Q3</t></is></c><c r="B9" t="inlineStr"><is><t>Helsinki</t></is></c><c r="C9"><v>23.1</v></c><c r="D9"><v>14</v></c><c r="E9"><v>16.12</v></c><c r="F9"><v>12.31</v></c></row><row r="10"><c r="A10" t="inlineStr"><is><t>2015Q3</t></is></c><c r="B10" t="inlineStr"><is><t>Turku</t></is></c><c r="C10"><v>15.32</v></c><c r="D10"><v>12.93</v></c><c r="E10"><v>10.65</v></c><c r="F10"><v>11.1</v></c></row><row r="11"><c r="A11" t="inlineStr"><is><t>Yksikkö: eur / m2</t></is></c></row><row r="12"><c r="A12" t="inlineStr"><is><t>Lähde: PxVisualizer-fi</t></is></c></row>
    </sheetData>
</worksheet>"
`;

exports[`Worksheet xml generation tests Should generate expexted xml output with the provided fixture with sv locale 1`] = `
"<?xml version="1.0" ?>
<worksheet xmlns="http://schemas.openxmlformats.org/spreadsheetml/2006/main" xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006" xmlns:mv="urn:schemas-microsoft-com:mac:vml" xmlns:mx="http://schemas.microsoft.com/office/mac/excel/2008/main" xmlns:r="http://schemas.openxmlformats.org/officeDocument/2006/relationships" xmlns:x14="http://schemas.microsoft.com/office/spreadsheetml/2009/9/main" xmlns:x14ac="http://schemas.microsoft.com/office/spreadsheetml/2009/9/ac" xmlns:xm="http://schemas.microsoft.com/office/excel/2006/main">
    <cols><col min="1" max="1" width="20"/>,<col min="2" max="2" width="20"/>,<col min="3" max="3" width="20"/>,<col min="4" max="4" width="20"/>,<col min="5" max="5" width="20"/>,<col min="6" max="6" width="20"/></cols>
    <sheetData>
        <row r="1"><c r="A1" t="inlineStr"><is><t>Kvadratmeterspris (eur/m2) 2015Q1-2015Q3 efter Område, Antal rum, Finansieringssätt</t></is></c></row><row r="2"><c r="C2" t="inlineStr"><is><t>Enrumslägenhet</t></is></c><c r="D2" t="inlineStr"><is><t>Enrumslägenhet</t></is></c><c r="E2" t="inlineStr"><is><t>Tvårumslägenhet</t></is></c><c r="F2" t="inlineStr"><is><t>Tvårumslägenhet</t></is></c></row><row r="3"><c r="C3" t="inlineStr"><is><t>Fri finansierad</t></is></c><c r="D3" t="inlineStr"><is><t>ARA</t></is></c><c r="E3" t="inlineStr"><is><t>Fri finansierad</t></is></c><c r="F3" t="inlineStr"><is><t>ARA</t></is></c></row><row r="4"><c r="C4" t="inlineStr"><is><t>Kvadratmeterspris (eur/m2)</t></is></c><c r="D4" t="inlineStr"><is><t>Kvadratmeterspris (eur/m2)</t></is></c><c r="E4" t="inlineStr"><is><t>Kvadratmeterspris (eur/m2)</t></is></c><c r="F4" t="inlineStr"><is><t>Kvadratmeterspris (eur/m2)</t></is></c></row><row r="5"><c r="A5" t="inlineStr"><is><t>2015Q1</t></is></c><c r="B5" t="inlineStr"><is><t>Helsingfors</t></is></c><c r="C5"><v>22.68</v></c><c r="D5"><v>13.8</v></c><c r="E5"><v>15.77</v></c><c r="F5"><v>12.09</v></c></row><row r="6"><c r="A6" t="inlineStr"><is><t>2015Q1</t></is></c><c r="B6" t="inlineStr"><is><t>Åbo</t></is></c><c r="C6"><v>15.15</v></c><c r="D6"><v>12.7</v></c><c r="E6"><v>10.5</v></c><c r="F6"><v>10.95</v></c></row><row r="7"><c r="A7" t="inlineStr"><is><t>2015Q2</t></is></c><c r="B7" t="inlineStr"><is><t>Helsingfors</t></is></c><c r="C7"><v>22.87</v></c><c r="D7"><v>13.95</v></c><c r="E7"><v>15.95</v></c><c r="F7"><v>12.23</v></c></row><row r="8"><c r="A8" t="inlineStr"><is><t>2015Q2</t></is></c><c r="B8" t="inlineStr"><is><t>Åbo</t></is></c><c r="C8"><v>15.18</v></c><c r="D8"><v>12.84</v></c><c r="E8"><v>10.57</v></c><c r="F8"><v>11.04</v></c></row><row r="9"><c r="A9" t="inlineStr"><is><t>2015Q3</t></is></c><c r="B9" t="inlineStr"><is><t>Helsingfors</t></is></c><c r="C9"><v>23.1</v></c><c r="D9"><v>14</v></c><c r="E9"><v>16.12</v></c><c r="F9"><v>12.31</v></c></row><row r="10"><c r="A10" t="inlineStr"><is><t>2015Q3</t></is></c><c r="B10" t="inlineStr"><is><t>Åbo</t></is></c><c r="C10"><v>15.32</v></c><c r="D10"><v>12.93</v></c><c r="E10"><v>10.65</v></c><c r="F10"><v>11.1</v></c></row><row r="11"><c r="A11" t="inlineStr"><is><t>Enhet: eur / m2</t></is></c></row><row r="12"><c r="A12" t="inlineStr"><is><t>Källa: PxVisualizer-sv</t></is></c></row>
    </sheetData>
</worksheet>"
`;
